### PR TITLE
Feature: extra mounts

### LINKED ISSUE #130: Allow custom volume mounts to aiidalab container
I've just started playing with `aiidalab-launch` and it works great! Much better experience than before.

However, it seems that I cannot currently utilize it since I need to have a custom volume mounted for my aiidalab application.* Would it be possible to add a profile setting to have custom volumes?

* my package depends on a quantum chemistry package ORCA, which is not open source and cannot be automatically installed. My current solution to this problem is to mount a directory with the ORCA binaries as a volume to the aiidalab container. I am all ears if there is a better approach. :-)

### PR DESCRIPTION
**Pull request by danielhollas**
Closes #130 

This needs tests and probably more error handling, but I'd like to have some eyes on this early to confirm if I am on the right track here.

CC @csadorf

### CODE REVIEW

File: aiidalab_launch/profile.py
Comment (danielhollas) [inline aiidalab_launch/profile.py:79]: I am using the Docker notation `source:target` here, but maybe there is a better solution?

File: aiidalab_launch/profile.py
Comment (danielhollas) [inline aiidalab_launch/profile.py:59]: I am new to this advanced `dataclass` business, but supposedly I cannot simply initialize to empty list here?

File: aiidalab_launch/util.py
Comment (danielhollas) [inline aiidalab_launch/util.py:130]: This is an unrelated mypy fix

File: aiidalab_launch/profile.py
Comment (csadorf) [inline aiidalab_launch/profile.py:82]: It could be a volume, no?

File: aiidalab_launch/profile.py
Comment (csadorf) [inline aiidalab_launch/profile.py:118]: Have a look at the output of `docker inspect $CONTAINER_ID`. You should be able to find the additional volumes there. If not, let me know I can try to investigate further.

File: aiidalab_launch/util.py
Comment (csadorf) [reply aiidalab_launch/util.py:130]: Could you elaborate a bit further on this? Does this occur with a specific mypy version? We are running [mypy checks as part of the CI](https://github.com/aiidalab/aiidalab-launch/blob/73fe854e525d1c0adfa1f92b1aa97842df5a5c16/.pre-commit-config.yaml#L37-L45) so that should have come up.

**Review by csadorf — COMMENTED** [review]
Thanks for the draft! I think this is a very good start. I've tried to address all of your comments and questions.

From a conceptual level I think this looks sound.

File: aiidalab_launch/profile.py
Comment (danielhollas) [reply aiidalab_launch/profile.py:59]: Thanks, that's what I suspected, so essentially it's the same gotcha as using an empty list as function optional parameter.

https://docs.python.org/3/library/dataclasses.html#mutable-default-values

File: aiidalab_launch/profile.py
Comment (danielhollas) [reply aiidalab_launch/profile.py:82]: Good catch, thanks. In `instance.py` we're determining whether we have a bind mount path or a volume name based on `Path.is_absolute()` so this check should work I guess

```python
       if source_path.is_absolute() and not source_path.is_dir():
            raise ValueError(f"Directory {source} does not exist")
```

By the way, I got very confused by this logic initially, and I didn't quite understood the difference between bind mounts and volume mounts. Perhaps the requirement for an absolute path for bind mounts should be mentioned in README / docs....

Comment (csadorf) [reply aiidalab_launch/profile.py:82]: Thanks for the hint. Would you want that additional documentation to target developers or users or both? And were you generally confused about the difference between bind mounts and volumes within docker or just in this context?

File: aiidalab_launch/__main__.py
Comment (danielhollas) [reply aiidalab_launch/__main__.py:272]: Agreed, thanks.

File: aiidalab_launch/util.py
Comment (danielhollas) [reply aiidalab_launch/util.py:130]: Here's the output from conda list

<details>

```
_libgcc_mutex             0.1                        main  
_openmp_mutex             5.1                       1_gnu  
aiidalab-launch           2022.1013                 dev_0    <develop>
appdirs                   1.4.4                    pypi_0    pypi
attrs                     21.4.0                   pypi_0    pypi
bumpver                   2021.1114                pypi_0    pypi
bzip2                     1.0.8                h7f98852_4    conda-forge
ca-certificates           2022.6.15            ha878542_0    conda-forge
cattrs                    1.10.0                   pypi_0    pypi
certifi                   2022.6.15        py39hf3d152e_0    conda-forge
cffi                      1.15.0           py39h4bc2ebd_0    conda-forge
cfgv                      3.3.1              pyhd8ed1ab_0    conda-forge
charset-normalizer        2.0.12                   pypi_0    pypi
click                     8.0.3                    pypi_0    pypi
click-spinner             0.1.10                   pypi_0    pypi
colorama                  0.4.4                    pypi_0    pypi
coverage                  6.4.1                    pypi_0    pypi
distlib                   0.3.4              pyhd8ed1ab_0    conda-forge
docker                    5.0.3                    pypi_0    pypi
dunamai                   1.7.0                    pypi_0    pypi
filelock                  3.7.1              pyhd8ed1ab_0    conda-forge
identify                  2.5.1              pyhd8ed1ab_0    conda-forge
idna                      3.3                      pypi_0    pypi
iniconfig                 1.1.1                    pypi_0    pypi
ld_impl_linux-64          2.38                 h1181459_1  
lexid                     2021.1006                pypi_0    pypi
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 11.2.0               h1234567_1  
libgomp                   11.2.0               h1234567_1  
libnsl                    2.0.0                h7f98852_0    conda-forge
libstdcxx-ng              11.2.0               h1234567_1  
libuuid                   2.32.1            h7f98852_1000    conda-forge
libzlib                   1.2.11            h166bdaf_1014    conda-forge
mypy                      0.910              pyhd3eb1b0_0  
mypy_extensions           0.4.3            py39h06a4308_1  
ncurses                   6.3                  h7f8727e_2  
nodeenv                   1.6.0              pyhd8ed1ab_0    conda-forge
openssl                   3.0.3                h166bdaf_0    conda-forge
packaging                 21.3                     pypi_0    pypi
pathlib2                  2.3.7.post1              pypi_0    pypi
pip                       21.2.4           py39h06a4308_0  
platformdirs              2.5.1              pyhd8ed1ab_0    conda-forge
pluggy                    1.0.0                    pypi_0    pypi
pre-commit                2.19.0           py39hf3d152e_0    conda-forge
psutil                    5.8.0            py39h27cfd23_1  
py                        1.11.0                   pypi_0    pypi
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.9                    pypi_0    pypi
pytest                    6.2.5                    pypi_0    pypi
pytest-asyncio            0.17.2                   pypi_0    pypi
pytest-cov                3.0.0                    pypi_0    pypi
python                    3.9.12          h2660328_1_cpython    conda-forge
python-dotenv             0.19.1                   pypi_0    pypi
python_abi                3.9                      2_cp39    conda-forge
pyyaml                    6.0              py39hb9d737c_4    conda-forge
readline                  8.1.2                h7f8727e_1  
requests                  2.26.0                   pypi_0    pypi
requests-cache            0.9.1                    pypi_0    pypi
responses                 0.18.0                   pypi_0    pypi
setuptools                61.2.0           py39h06a4308_0  
six                       1.16.0             pyh6c4a22f_0    conda-forge
sqlite                    3.38.5               h4ff8645_0    conda-forge
tabulate                  0.8.9                    pypi_0    pypi
tk                        8.6.12               h27826a3_0    conda-forge
toml                      0.10.2             pyhd3eb1b0_0  
tomli                     2.0.1                    pypi_0    pypi
types-click-spinner       0.1.11                   pypi_0    pypi
types-pyyaml              6.0.8                    pypi_0    pypi
types-redis               4.2.7                    pypi_0    pypi
types-requests            2.27.30                  pypi_0    pypi
types-tabulate            0.8.9                    pypi_0    pypi
types-toml                0.10.7                   pypi_0    pypi
types-ujson               5.3.0                    pypi_0    pypi
types-urllib3             1.26.15                  pypi_0    pypi
typing_extensions         4.1.1              pyh06a4308_0  
tzdata                    2022a                hda174b7_0  
ukkonen                   1.0.1            py39hf939315_2    conda-forge
url-normalize             1.4.3                    pypi_0    pypi
urllib3                   1.26.9                   pypi_0    pypi
virtualenv                20.14.1          py39hf3d152e_0    conda-forge
websocket-client          1.3.2                    pypi_0    pypi
wheel                     0.37.1             pyhd3eb1b0_0  
xz                        5.2.5                h7f8727e_1  
yaml                      0.2.5                h7f98852_2    conda-forge
zlib                      1.2.11            h166bdaf_1014    conda-forge
```
</details>

File: aiidalab_launch/instance.py
Comment (danielhollas) [inline aiidalab_launch/instance.py:144]: This is a bugfix unrelated to the changes in this PR. `docker_mount_for` returns Path for bind mounts, while `mount["Source"]` is always str so this check would never be true even though the configuration did not change, and the user would get spurious warnings. LMK if you want me submit a separate PR for this.

File: aiidalab_launch/profile.py
Comment (danielhollas) [reply aiidalab_launch/profile.py]: I somehow thought that volumes are always tied to existing paths (bind), did not know that Docker can manage them internally. It was especially confusing when I first started playing `aiidalab-launch` in default settings, and was very confused that the container was not using my preexisting aiidalab folder in my home folder.

I can try to write something out to README.md for this PR, since I guess we need to document the new functionality anyways.

File: aiidalab_launch/util.py
Comment (csadorf) [reply aiidalab_launch/util.py:130]: We are using mypy version 0.961, not 0.910 . I am doubtful that this is the issue, but would be at least worth to check. Are the pre-commit hooks passing for you?

File: aiidalab_launch/profile.py
Comment (csadorf) [reply aiidalab_launch/profile.py]: Yes, if you could improve the documentation that would be very much appreciated. Could be as part of this PR.

File: aiidalab_launch/instance.py
Comment (csadorf) [reply aiidalab_launch/instance.py:144]: I would be ideal to fix this bug in a separate PR. There we could also try to add a simple test that catches this bug. This function is tested [pretty extensively](https://github.com/aiidalab/aiidalab-launch/blob/73fe854e525d1c0adfa1f92b1aa97842df5a5c16/tests/test_instance.py#L78) – so I would think that the undesired behavior should show up there.

**Comment by csadorf** [comment]
> @csadorf Thank you for the initial review and detailed comments. I implemented your suggestions, added a readonly option ("source:target:readonly") and added determination of extra mounts for the container (needed for determining whether configuration changed underneath a running container).

That's great, very much appreciated. Since we are essentially supporting the ["short-syntax" from docker-compose](https://docs.docker.com/compose/compose-file/compose-file-v3/#volumes), I would recommend that we use "rw" (implied) and "ro" to indicate read-write and read-only mounts.

> I still need to add tests. I am not very familiar with pytest yet, so pointers welcome.

All tests are implemented in the [`tests/`](https://github.com/aiidalab/aiidalab-launch/tree/main/tests) directory, we use a [hierarchy of test fixtures](https://github.com/aiidalab/aiidalab-launch/blob/main/tests/conftest.py) to isolate the tests from the actual user environment (we do not want to mess with the user's actual aiidalab containers and volumes). Since all of that is already taking care of, you don't typically have to worry about that too much and can just pattern-match the existing tests. Maybe I can write a short summary of how to develop tests and add it to the README?

> EDIT: Just realized I also need to write a migration to convert old config files to include the new option, perhaps in
> 
> https://github.com/aiidalab/aiidalab-launch/blob/73fe854e525d1c0adfa1f92b1aa97842df5a5c16/aiidalab_launch/application_state.py#L66
> 
> ?

Yes, I think it might be as simple as always updating the config version instead of only updating it when there is no previous version stored; something like this:

```python
        # Update config version if necessary
        if self.config.version != str(parse(__version__))
            self.config.version = str(parse(__version__))
            config_changed = True
```
The `save_config()` function should write the whole config (including defaults – to be verified).

File: aiidalab_launch/instance.py
Comment (danielhollas) [reply aiidalab_launch/instance.py:144]: @csadorf submitted a new PR #134 and added a test. I will revert this change on this PR and rebase/merge with main after #134 is merged.

File: aiidalab_launch/util.py
Comment (danielhollas) [reply aiidalab_launch/util.py:130]: Pre-commit hooks are passing. I have updated to 0.961 but the issue persisted. But I am happy to revert this at the end (keeping for now so mypy output is clean during development).

**Comment by csadorf** [comment]
@danielhollas Please re-request my review when you want me to have another detailed look at this. Thank you!

**Comment by danielhollas** [comment]
> Since we are essentially supporting the ["short-syntax" from docker-compose](https://docs.docker.com/compose/compose-file/compose-file-v3/#volumes), I would recommend that we use "rw" (implied) and "ro" to indicate read-write and read-only mounts.
Huh. So I was originally looking at these [this Docker docs](https://docs.docker.com/storage/volumes/#choose-the--v-or---mount-flag) which make it seems like the third argument can function as a list of options. But I tested that it doesn't working with my Docker version so I guess those Docs are outdated. So I went with your suggestion for `ro|rw`, with the default being `rw` and being explicit (e.g. if user does not provide it, we will autocomplete it). 

> Yes, I think it might be as simple as always updating the config version instead of only updating it when there is no previous version stored; something like this:
```python
        # Update config version if necessary
        if self.config.version != str(parse(__version__))
            self.config.version = str(parse(__version__))
            config_changed = True
```
Yes, this indeed works and solved the migration issue.

I added a bunch of tests and apart from the printing in `__main__.py` everything is covered according to Codecov. 
If this looks good I can add some Docs, but unfortunately I will not have too much extra time to work on this in the next two weeks (conference time). I will be using this actively so am happy to follow up if I discover any issues once this is released.

File: aiidalab_launch/profile.py
Comment (danielhollas) [inline aiidalab_launch/profile.py:99]: I couldn't find this in any Docs, but Docker actually prints this regex in an error message. :-)
This validation could be added to the home mount as well.

File: tests/test_cli.py
Comment (danielhollas) [inline tests/test_cli.py:218]: I tried a bunch of stuff but couldn't make this work...

File: aiidalab_launch/profile.py
Comment (csadorf) [reply aiidalab_launch/profile.py:99]: Nice find! 👍

File: tests/conftest.py
Comment (csadorf) [inline tests/conftest.py:137]: What happens if the volume does not exist or was already removed?

File: aiidalab_launch/profile.py
Comment (csadorf) [reply aiidalab_launch/profile.py:99]: Can we add a `valid_volume_name` function to the `util` module and then use it for the other volume names as well?

File: tests/test_profile.py
Comment (csadorf) [inline tests/test_profile.py:13]:
```suggestion
    "{path}:/opt/test",
```

File: tests/test_cli.py
Comment (csadorf) [reply tests/test_cli.py:218]: I think manipulating the `instance` variable like this may not have any effect on the profile that is picked up by the CLI runner. You could check that by printing the output of `runner.involke(cli.cli, ["profiles", "show", "default"])`.

I can envision two ways to address this:
1. Create an additional fixture named `use_extra_volumes` or so that actually changes the configuration when requested.
2. Just configure all tests to use extra_volumes.

I don't like the 2nd approach, because it means that we are moving further away from the user default config in our tests, which reduces our test effectiveness, however it might be a good way to debug this problem.

**Review by csadorf — COMMENTED** [review]
Looks good overall! I have left a few comments and suggestions and my recommendation on how we might be able to address that test issue.

File: tests/test_cli.py
Comment (danielhollas) [reply tests/test_cli.py]: I have already tried the second approach by modifying the Profile fixture, but that didn't work. I think we would need to go deeper and modify the config fixture.

I would need to think more about how to create `use_extra_volumes` fixture.

File: aiidalab_launch/profile.py
Comment (danielhollas) [reply aiidalab_launch/profile.py]: I have decided to keep this function in profile instead of util, because almost the same regex already existed for validating the profile names so it made sense to me to have both of them at the same place.

File: aiidalab_launch/profile.py
Comment (danielhollas) [inline aiidalab_launch/profile.py:154]: This code flaky unfortunately, we're not guaranteed to get the extra mounts in the same order as they appeared in the original Profile. Because of this, the test in test_instance.py is flaky. https://github.com/aiidalab/aiidalab-launch/runs/7045495764?check_suite_focus=true

Not sure how to deal with this. Maybe we need to pass the expected extra_mounts as an optional parameter here to match the order. We could also specialize the `__eq__` method for Profile, but that seems like an awful lot of work.

**Comment by danielhollas** [comment]
@csadorf thanks, I have implemented the suggestions. Still not quite sure how to deal with the test_cli test, but I have unfortunately found a more pressing issue with the `Profile.from_container` which results in the test flakiness, see this comment for more details. 

https://github.com/aiidalab/aiidalab-launch/pull/132/files#r906282515

File: aiidalab_launch/profile.py
Comment (danielhollas) [reply aiidalab_launch/profile.py:157]: Okay, I think I solved by converting extra_mounts to Set(). Wasn't even that difficult a change, although I did have to patch the TOML loading.

File: README.md
Comment (csadorf) [inline README.md:47]:
```suggestion
### Data Management

```

File: README.md
Comment (csadorf) [inline README.md:50]:
```suggestion
You can also provide an absolute path to the AiiDAlab home directory on the host system
```

File: README.md
Comment (csadorf) [inline README.md:52]:
```suggestion
If this directory does not exist, AiiDAlab launch will try to create it on startup.
```
"AiiDAlab launch" referring to the software by name
`aiidalab-launch` when referring to the executable 😉

File: README.md
Comment (csadorf) [inline README.md:59]: Can you mention the conda volume and add one example?

**Review by csadorf — CHANGES_REQUESTED** [review]
Looks like this is almost ready! 😃 I've tested it locally and everything seems to be working as expected.

I've identified one bug and proposed a fix as well as merged a unit test that catches it in #136. Maybe you can rebase this branch so that we can test against it?

I have a few more minor change requests, but should all be easily addressable.

One last thing, I noticed that the volume is root-owned by default. That's expected, however should we try to address that somehow? I think we should not mess with ownership and permissions by default, but maybe it could be an option? Please let me know what you suggest.

File: aiidalab_launch/config.py
Comment (danielhollas) [reply aiidalab_launch/config.py]: Thanks for catching this, I've rebased on main and implemented your fix.

File: README.md
Comment (danielhollas) [reply README.md:55]: I've added an example, but I don't think I understand the conda volume (its purpose and how to work with it) well enough to add it. Feel free to push a commit with your suggestion.

I know that it is mounted in `/home/aiida/.conda` but I did not figure out how would I actually use it, since `conda install` still points to `/opt/`.

**Comment by danielhollas** [comment]
@csadorf thanks for shepherding this PR. :-) I've rebased on main and implemented your bugfix, thanks for catching that.

> One last thing, I noticed that the volume is root-owned by default. That's expected, however should we try to address that somehow? I think we should not mess with ownership and permissions by default, but maybe it could be an option? Please let me know what you suggest.
I am confused, what exactly do you mean by root-owned? In my testing the extra mounts are owned by the aiida user in the container, same as the home_mount and conda volume.

File: README.md
Comment (csadorf) [reply README.md:55]: Ok, no problem. I will see if I can clarify that.

**Comment by csadorf** [comment]
> @csadorf thanks for shepherding this PR. :-) I've rebased on main and implemented your bugfix, thanks for catching that.
> 
> > One last thing, I noticed that the volume is root-owned by default. That's expected, however should we try to address that somehow? I think we should not mess with ownership and permissions by default, but maybe it could be an option? Please let me know what you suggest.
> 
> I am confused, what exactly do you mean by root-owned? In my testing the extra mounts are owned by the aiida user in the container, same as the home_mount and conda volume.

When mounting a volume into a container the UID and GID are not changed. That means for a bind mount that the ownership of the source directory are maintained and you may not have access to the directory if the user in the container has a different UID. [This blog article](https://medium.com/@mccode/understanding-how-uid-and-gid-work-in-docker-containers-c37a01d01cf) might provide additional insight.

For aiidalab-launch we [currently use the "aiida" user](https://github.com/aiidalab/aiidalab-launch/blob/891d74e7d63e720d8e2f56e8c1077e6a7d15f02e/aiidalab_launch/profile.py#L56), which has a UID and GID of 1000. That's the standard user id for the first user on most distributions so in general this will work just fine. On my system it is the UID of the root user which means I don't automatically have access to it. Long story short, since you are not encountering this problem and we can assume that most people won't, let's maybe hold off on implementing a "solution" to it since it could be easily done in a separate change set.

**Comment by danielhollas** [comment]
@csadorf awesome, thank you for the final touches, lgtm.

**Comment by danielhollas** [comment]
@csadorf thanks for merging! :rocket:  Would you be able to create a new release today or tomorrow?

**Comment by csadorf** [comment]
> @csadorf thanks for merging! 🚀 Would you be able to create a new release today or tomorrow?

[Done](https://github.com/aiidalab/aiidalab-launch/releases/tag/v2022.1014).

https://pypi.org/project/aiidalab-launch/2022.1014/

Let me know if there is any issue, I can make a new release very quickly.

**Comment by danielhollas** [comment]
@csadorf excellent, thank you very much. I have just tested it and it seems to work. btw; on another computer I am indeed running into the permission issue with the mounted volume that you mentioned above. I'll take a closer look at that medium post and think about what could be done (although it might take me some time before I get to it). For now the simple fix is to chmod add access to all users, which in my case is okay but not a great solution in general.

**Pull request merged**
